proposals/nnnn-distributed-actors.md

```diff
@@ -0,0 +1,1620 @@
+
+As a consequence of the request-response nature of distributed methods, `inout` parameters are not supported. While subscripts are similar to methods, they are not allowed for a distributed actor. A subscript's usefulness is strongly limited by both their lack of support for being `distributed` (e.g., could only support read-only subscripts, because no coroutine-style accessors) and their lightweight syntax can lead to the same problems as properties.
+
+
+### Distributed Actor lifecycle
+
+The lifecycle of a distributed actor is important to consider, because it is a key part of how location transparency is achieved.
+
+#### Local Initialization
+
+All user-defined designated initializers of a distributed actor can be collectively referred to as "local" initializers. All local initializers of a distributed actor must accept exactly one `ActorTransport` parameter. This `ActorTransport` parameter is implicitly used to fulfil an important contract with the transport and the actor instance itself. Namely, all distributed actor instances must be assigned an `ActorIdentity`, which is provided by the transport. In addition, once fully initialized, the actor must inform the transport that it is ready to receive messages. Conceptually, these steps can be described as:
+
+```swift
+distributed actor DA {
+  init(..., transport: ActorTransport) {
+    // Step 1: Try to set-up the transport and identity.
+    self._transport = transport
+    self._id = AnyActorIdentity(transport.assignIdentity(Self.self))
+    
+    // Step 2: User code is executed, which may fail, access the identity, etc.
+    
+    // Step 3: The initializer did not fail; notify the transport.
+    transport.actorReady(self)
+  }
+}
+```
+
+If no user-defined designated initializer is provided, a default initializer is synthesized that requires a named parameter `transport` of type `ActorTransport`, i.e., `init(transport:)`. The access level of the synthesized init is the same as the distributed actor declaration (i.e. for an internal distributed actor, the init is internal, and public for a public one). As with other nominal types, this initializer is only synthesized by the compiler if all stored properties can be default-initialized.
+
+```swift
+distributed actor DA {
+  // synthesized: init(transport: ActorTransport) { ... }
+}
+```
+
+#### Remote Resolution
+
+There is no representation of a remote initializer for a distributed actor, because a remote instance has *conceptually* already been initialized. Thanks to location transparency, a remote instance is not even guaranteed to exist until it is actually needed. This is an important implementation detail that enables efficient implementations of distributed systems. To make this more concrete, consider a simplified version of the static `resolve` function that is synthesized for all distribtued actors:
```

Comment:
```suggestion
There is no representation of a remote initializer for a distributed actor, because a remote instance has *conceptually* already been initialized. Thanks to location transparency, a remote instance is not even guaranteed to exist until it is actually needed. This is an important implementation detail that enables efficient implementations of distributed systems. To make this more concrete, consider a simplified version of the static `resolve` function that is synthesized for all distributed actors:
```